### PR TITLE
[core] Simplify options handling [Part 1]

### LINKED ISSUE #23029: [core][Bug] Ray DAG does not capture options correctly
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### Issue Severity

Low: It annoys or frustrates me for a moment.

### What happened + What you expected to happen

Ray DAG fails to capture the options of some ray tasks, or the options it get differs from what we expect.

### Versions / Dependencies

Latest Ray

### Reproduction script

```python
@ray.remote(num_gpus=100)
def foo(): pass

foo._bind().get_options() # {}

@ray.remote
def bar(): pass

bar.options(num_gpus=100)._bind().get_options() # {'num_returns': None, 'num_cpus': None. 'num_gpus': 100 ....}

```

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

**Changes to API behavior**

1. (non-breaking change) All options in `.options()` are supported in `@ray.remote`. Semantically `@ray.remote` provides default options for a remote function/actor. Previously some are supported, but some are not, this is confusing.
2. (breaking, but very unlikely to affect users) Options in `.options` overrides the options in `@ray.remote`. This is the behavior in the docstring, but due to some implementation flaws, the real situation could be confusing. For example, `ray.remote(num_cpus=2).options(num_cpus=None) => ray.remote(num_cpus=2)`, `ray.remote(placement_group=None).options(placement_group="default") => ray.remote(placement_group=None)`. Now it becomes `ray.remote(num_cpus=2).options(num_cpus=None) => ray.remote(num_cpus=None)`

all changes are confirmed in #api-changes

---

I was about to fix https://github.com/ray-project/ray/issues/23029, but then I find options handling was totally a mess. So I decided to cleanup the code related to options handling first. I decide to cleanup part of them so it is easier for review.

The design principle I follow:
* Use a centralized schema/metadata to statically define options names, types, constraints and default values.
* Fail early. If we can know an option is invalid, throw an error earlier. This improves user experience.
* Explicit error types. If it is an option type error, throw `TypeError`. If it violates constraints, throw `ValueError`. (previously it is a mixture of `AssertionError`, `TypeError`, `ValueError` and the error type does not match the cause).
* Keep the user input options. This helps debugging and merging options (otherwise you have to merge processed options and new options).
* Be explicit about options that could only be used in `.options()`
* Uniform error message - easier to catch and parse
* Cross-language friendly - use constants/basic types for default values.

The known design errors in the original code:

* The type and range of `max_pending_calls` is not checked in client API.
* Mixed use of `dict.get()` and "if not in dict return None"
* Default value assignment in various levels, for example, `concurrency_groups` was assigned in `remote()`, `num_cpus` was assigned in `RemoteFunction`
* Value check in different levels, some are duplicated, some are missing in certain code path.
* Incorrect error type. For example, `TypeError` for unexpected keywords in client API.
* Duplicated code. Large amount of code duplication in client API and non-client API for parsing arguments.
* Some options like `max_pending_calls` can only be used in `.options()` but not in `@remote`. Need to be more clear in code.
* ...

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by suquark** [comment]
@rkooo567 ready for review now 🙂

File: python/ray/tests/test_client.py
Comment (suquark) [inline python/ray/tests/test_client.py:484]: now we raise the error earlier, so you expect an error before `ray.get()`. this helps catching errors when defining the remote function.

File: python/ray/tests/test_basic.py
Comment (suquark) [inline python/ray/tests/test_basic.py:183]: test is enhanced

File: python/ray/actor.py
Comment (suquark) [inline python/ray/actor.py:359]: these are possible inconsistency in our code. I will address them a later PRs.

File: python/ray/actor.py
Comment (suquark) [inline python/ray/actor.py:584]: This is likely inconsistency in our current API. Will address it in later PRs.

File: python/ray/actor.py
Comment (suquark) [inline python/ray/actor.py:1324]: another inconsistency: we only check max_int64 for a single integer. we should apply it to all later. also I think we should raise an error instead of rounding it quietly

**Comment by jjyao** [comment]
I like more reds than greens! I'll review it today.

File: python/ray/util/options.py
Comment (jjyao) [inline python/ray/util/options.py:167]: I feel we should just have the same error message for both cases: Invalid option keyword {k} for remote functions, Valid ones are {task_options}

File: python/ray/util/options.py
Comment (jjyao) [inline python/ray/util/options.py:175]: tasks -> actors

File: python/ray/worker.py
Comment (jjyao) [inline python/ray/worker.py:2106]: Why do we need to do this?

File: python/ray/util/client/options.py
Comment (jjyao) [inline python/ray/util/client/options.py:17]: Should we validate task and actor options separately since they have their own unique set of options?

File: python/ray/remote_function.py
Comment (jjyao) [inline python/ray/remote_function.py:178]: I think it's clearer to implement our own merge function that considers default value: options.merge(dict1, dict2).

Comment (jjyao) [inline python/ray/remote_function.py:178]: nit:  _options and _option_dict sound similar, maybe name then as _options_from_remote and _options_from_options?

File: python/ray/remote_function.py
Comment (jjyao) [inline python/ray/remote_function.py:90]: Actually not just tests, for example `_max_calls` is used in function_manager.py

**Review by jjyao — COMMENTED** [review]
Overall direction lgtm

File: python/ray/util/options.py
Comment (suquark) [reply python/ray/util/options.py:1]: I put it under `util` because ray client also put related/similar code under `utils`.

File: python/ray/util/options.py
Comment (suquark) [reply python/ray/util/options.py:22]: Currently only `max_calls` is `remote_only`, I am not sure if it is over design.

File: python/ray/util/options.py
Comment (suquark) [reply python/ray/util/options.py:20]: I think usually they are separate to avoid confusion. E.g., if a default value is a lambda function.

File: python/ray/util/options.py
Comment (suquark) [reply python/ray/util/options.py:79]: surprisingly, they are integers we I look into code and test cases

File: python/ray/util/options.py
Comment (suquark) [reply python/ray/util/options.py:95]: here we can use `None` as default value, but users should not use `None` as the input, because `None` is not a valid `runtime_env`

File: python/ray/util/options.py
Comment (suquark) [reply python/ray/util/options.py:97]: yes, we should later delete the one in `_raylet.pyx`. Type check should be forwarded.

File: python/ray/util/options.py
Comment (suquark) [reply python/ray/util/options.py:167]: Hmmm, I think this is the original design, because users may use options of actors for remote functions. But yes we can just use a uniformed message, the original one could be more confusing.

File: python/ray/worker.py
Comment (suquark) [reply python/ray/worker.py:2106]: This is something messy in the original code: there are two (or three) sets of default values: in some places `None` is the default value to be overrided, in some places `None` is the default value you should not override (`None` has some meanings), in other places the default value is number or string. I do not want to change too much code, so `user_input_filter` is just a workaround to unify them.

File: python/ray/util/client/options.py
Comment (suquark) [reply python/ray/util/client/options.py:17]: That's for the future, I do not want to further increase the LoC, otherwise it takes much longer time to merge.

File: python/ray/util/options.py
Comment (rkooo567) [reply python/ray/util/options.py:1]: Feel like both should go under /_private unless they are meant to be exposed to users (which I believe it doesn't).

File: python/ray/util/options.py
Comment (rkooo567) [reply python/ray/util/options.py:64]: Also the type annotation for inputs

File: python/ray/util/options.py
Comment (rkooo567) [inline python/ray/util/options.py:24]: Maybe we need unit tests for this file?

File: python/ray/util/options.py
Comment (rkooo567) [inline python/ray/util/options.py:190]: This looks like it is always used with `validate_actor_options ` or `validate_task_options ` except one place.

Is there any way to make this as a private function used in `validate_actor_options` & `validate_task_options`? Or would you like to make them a const method?

File: python/ray/util/options.py
Comment (jjyao) [reply python/ray/util/options.py:22]: I feel this is cleaner conceptually: an option can be remote only, options only and both. We don't need to make exceptions somewhere else for max_calls.

File: python/ray/util/options.py
Comment (jjyao) [reply python/ray/util/options.py:79]: anyway, this option will be removed in ray 2.0.

File: python/ray/util/options.py
Comment (jjyao) [reply python/ray/util/options.py:97]: What about value checks?

File: python/ray/remote_function.py
Comment (jjyao) [reply python/ray/remote_function.py]: How is this resolved?

Comment (jjyao) [reply python/ray/remote_function.py]: How is this resolved, seems the code is not changed. (I'm totally fine with rejecting the comment :), just want to see the reason.

File: python/ray/actor.py
Comment (jovany-wang) [reply python/ray/actor.py:584]: Yes. It only supports `.remote` according to the doc.

File: python/ray/remote_function.py
Comment (suquark) [reply python/ray/remote_function.py]: this is removed

Comment (suquark) [reply python/ray/remote_function.py]: this is removed

File: python/ray/util/options.py
Comment (suquark) [reply python/ray/util/options.py]: I further checked the code, it seems that all options allowed in `.options()` is equally valid for `@ray.remote`, it is more likely that we did not support it because people did not add update `@ray.remote` to sync with new options. I will solidify this choice by asking people related to API changes.

Also only `max_calls` and `max_concurrency` are `@ray.remote` only, so I decide remove `options_only` and `remote_only` at the same time.

File: python/ray/util/options.py
Comment (suquark) [reply python/ray/util/options.py]: fixed

File: python/ray/util/options.py
Comment (suquark) [reply python/ray/util/options.py]: both value and type checks should be done early.

File: python/ray/remote_function.py
Comment (suquark) [reply python/ray/remote_function.py]: I cleaned up this

File: python/ray/remote_function.py
Comment (suquark) [reply python/ray/remote_function.py]: I raised an exception when validating options

File: python/ray/util/options.py
Comment (suquark) [reply python/ray/util/options.py]: I removed this function. It is required as a workaround to change less code. Now I cleaned up more code and this is no longer required.

File: python/ray/util/options.py
Comment (suquark) [reply python/ray/util/options.py]: After looking into other options, `default_value_factory` is only used by `concurrency_groups`. I decide to remove it.

**Comment by suquark** [comment]
I cleaned up more code without significantly increase the LoC. This addresses some confusion.

File: python/ray/remote_function.py
Comment (suquark) [inline python/ray/remote_function.py:333]: this is how ray actor handles names. here we unify the behavior

**Comment by suquark** [comment]
@jjyao @rkooo567 ready for review again

**Comment by suquark** [comment]
also cc @iycheng , because after this PR, I can add the `metadata` keyword option easier.

File: python/ray/actor.py
Comment (jjyao) [inline python/ray/actor.py:533]: Seems `_option_dict` is duplicate of what's in `meta`, are we going to unify them?

Comment (jjyao) [inline python/ray/actor.py:533]: IIUC, this `self._option_dict` is for options specified through .remote() right? Maybe it's clear to name them `options_via_remote_fn` and `options_via_options_fn` (or something similar) instead of old and new?

File: python/ray/actor.py
Comment (jjyao) [inline python/ray/actor.py:588]: The `actor_options` here is already merged options? If so we probably don't need the extra merge that's happening in this method (maybe you already planned it as a follow-up).

**Review by jjyao — COMMENTED** [review]
lgtm

File: python/ray/actor.py
Comment (suquark) [reply python/ray/actor.py:533]: yes! for later PRs

File: python/ray/actor.py
Comment (suquark) [reply python/ray/actor.py:588]: are you referring to

```python
        # fill actor required options
        for k, v in options.actor_options.items():
            actor_options[k] = actor_options.get(k, v.get_default_value())
```
?

here we are not merging it again, just fulfilling missing options. we did similar stuff to `__ray_metadata__`, but unfortunately (and annoyingly) `__ray_metadata__` lacks a few fields, so we still need to fill missing options again. I decide to clean it up later.

**Comment by suquark** [comment]
According to the discussion in slack (api-changes), the conclusion is that we should support all `.option()` keywords in `@ray.remote`, this keeps our code and semantic consistent. This PR already implemented this.

File: python/ray/actor.py
Comment (jjyao) [reply python/ray/actor.py:588]: We have code like `scheduling_strategy = scheduling_strategy or meta.scheduling_strategy` which I think is no longer needed?

Comment (suquark) [reply python/ray/actor.py:588]: yes, that's true. let me clean these fields

File: python/ray/remote_function.py
Comment (jjyao) [inline python/ray/remote_function.py:145]: nit: Could we also rename `self._option_dict` to `self._options_from_remote_decorator`? It's a bit longer but I feel it's clearer?

File: python/ray/remote_function.py
Comment (jjyao) [inline python/ray/remote_function.py:149]: Actually now I'm thinking of it, seems it does not exactly match the current merge logic?

For example
@ray.remote(placement_group=pg)
def func():
   pass

func.option(placement_group="default").remote()

The old behavior is that we will use `pg` as the placement group, but the new logic will just use "default"?

File: python/ray/actor.py
Comment (iycheng) [inline python/ray/actor.py:684]: I feel here a dataclass seems better.

Comment (suquark) [reply python/ray/actor.py:684]: yes, this is just a temporary solution.

File: python/ray/remote_function.py
Comment (suquark) [reply python/ray/remote_function.py]: I decide to use `default_options` because not all of them come from `ray.remote`, especially for actors

File: python/ray/actor.py
Comment (suquark) [reply python/ray/actor.py:695]: I add a TODO for it

File: python/ray/remote_function.py
Comment (suquark) [reply python/ray/remote_function.py]: I looked into the code and docs, "default" is not supposed to be used by users, it is just a placeholder representing "not specified in .options()". This is because `None` cannot be used to indicating the missing of `placement_group` field, it is a valid option.

So the logic is still consistent with the original code, just the original code process missing keywords in a weird way

File: python/ray/remote_function.py
Comment (jjyao) [reply python/ray/remote_function.py]: oh, could you give me an example?

File: python/ray/remote_function.py
Comment (jjyao) [reply python/ray/remote_function.py]: It's actually not just about "default" pg: the original code cannot differentiate between user explicitly setting an option to default (e.g. None, "default") and user doesn't set the option, so it will try to use the option set via remote() if the options() one has the default value. With the new code, IIUC, we can now differentiate these two cases by the existence of the option key in the dict and will use the options() one as long as it's set by the user explicitly regardless what the value is. Although normally, users don't set the option to default value explicitly but it's still possible that some users do `f.options(some_option=None).remote()` and expect to use the value of some_option set via remote().

**Comment by iycheng** [comment]
I like the direciton of this PR. I have some concerns here.

Will the doc still work here?

For example, right now we have

```
>>> import ray
ray.>>> ray.remote
<function remote at 0x7fba6a4b3dc0>
>>> help(ray.remote)

>>> def f(): pass
...
>>> help(ray.remote(f).options)

Help on method options in module ray.remote_function:

options(args=None, kwargs=None, num_returns=None, num_cpus=None, num_gpus=None, memory=None, object_store_memory=None, accelerator_type=None, resources=None, max_retries=None, retry_exceptions=None, placement_group='default', placement_group_bundle_index=-1, placement_group_capture_child_tasks=None, runtime_env=None, name='', scheduling_strategy: Union[NoneType, str, ray.util.scheduling_strategies.PlacementGroupSchedulingStrategy] = None) method of ray.remote_function.RemoteFunction instance
```

Another one is that could we model the options in dataclass and move the type checking and (*args, **kwargs) converting there? 


NIT: The last one is not a concern just some confusions when I'm reading the code. Maybe when referring to options library, let's use ray._priavate.options or just rename the name to ray_options. Because sometimes, I got confused with .options in remote_function/actors and .options in the code.

File: python/ray/remote_function.py
Comment (suquark) [reply python/ray/remote_function.py]: I disagree with this kind of usage. It is meaningless to set a value to `None/"default"` just for reusing the original value - instead you avoid setting the value in `.options()`.

So to me in the original code, this is just a way to detect missing of a field instead of "reusing the original option".

File: python/ray/util/options.py
Comment (suquark) [reply python/ray/util/options.py]: test added

**Comment by suquark** [comment]
The CI failure is due to https://github.com/ray-project/ray/issues/23616

File: python/ray/remote_function.py
Comment (suquark) [reply python/ray/remote_function.py]: according to the discussion in slack, we keep the design of this PR

**Comment by suquark** [comment]
@jjyao @iycheng looks like all issues have been resolved

> I like the direciton of this PR. I think it's going to be more and more important in the short term since we are doing unification of workflow and ray.
> 
> Yah, I have some concerns here.
> 
> Will the doc still work here?
> 
> For example, right now we have
> 
> ```
> >>> import ray
> ray.>>> ray.remote
> <function remote at 0x7fba6a4b3dc0>
> >>> help(ray.remote)
> 
> >>> def f(): pass
> ...
> >>> help(ray.remote(f).options)
> 
> Help on method options in module ray.remote_function:
> 
> options(args=None, kwargs=None, num_returns=None, num_cpus=None, num_gpus=None, memory=None, object_store_memory=None, accelerator_type=None, resources=None, max_retries=None, retry_exceptions=None, placement_group='default', placement_group_bundle_index=-1, placement_group_capture_child_tasks=None, runtime_env=None, name='', scheduling_strategy: Union[NoneType, str, ray.util.scheduling_strategies.PlacementGroupSchedulingStrategy] = None) method of ray.remote_function.RemoteFunction instance
> ```
> 
> Another one is that could we model the options in dataclass and move the type checking and (*args, **kwargs) converting there?
> 
> NIT: The last one is not a concern just some confusions when I'm reading the code. Maybe when referring to options library, let's use ray._priavate.options or just rename the name to ray_options. Because sometimes, I got confused with .options in remote_function/actors and .options in the code.

The docstring still works. I updated the doc of `@ray.remote`. The dataclass stuff will be done in later PRs.

**Comment by suquark** [comment]
@jjyao @iycheng @pcmoritz take a review (again)?

**Review by fishbone — APPROVED** [review]
LGTM. Please wait for another stamp before merging since this PR is big.

**Review by jjyao — APPROVED** [review]
Since it's api-change, we need @pcmoritz's approval as well.

Great work on the cleanup!

**Review by rkooo567 — APPROVED** [review]
One last comment. I will also wait for APi approval before merging the PR!

File: python/ray/_private/ray_option_utils.py
Comment (suquark) [reply python/ray/_private/ray_option_utils.py:134]: oh. this is just a comment to explain the code. the behavior is aligned with the original one

**Review by pcmoritz — APPROVED** [review]
Thanks a lot for doing this :)

**Pull request merged**
